packages/tpetra/core/example/BlockCrs/driver.cpp

```diff
@@ -0,0 +1,536 @@
+
+      Kokkos::parallel_for
+        (num_owned_elements, KOKKOS_LAMBDA(const LO row) {
+          for (LO loc=rowptr(row);loc<rowptr(row+1);++loc) {
+            const GO gid_row = mesh_gids(row);
+            const GO gid_col = mesh_gids(colidx(loc));
+            
+            LO i0, j0, k0, i1, j1, k1;
+            sb.idx_to_ijk(gid_row, i0, j0, k0);
+            sb.idx_to_ijk(gid_col, i1, j1, k1);
+            
+            const LO diff_i = i0 - i1;
+            const LO diff_j = j0 - j1;
+            const LO diff_k = j0 - k1;
+            
+            auto block = blocks(loc);
+            for (LO l0=0;l0<blocksize;++l0) 
+              for (LO l1=0;l1<blocksize;++l1) 
+                block(l0, l1) = get_block_crs_entry<value_type>(i0, j0, k0,
+                                                                diff_i, diff_j, diff_k,
+                                                                l0, l1);
+          }
+        });
+    }
+    {
+      RCP<TimeMonitor> timerBlockCrsFillComplete 
+        = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("3) BlockCrsMatrix FillComplete - currently do nothing")));
+      // this function does not exsit 
+      //A_bcrs->fillComplete();
+    }
+#if PRINT_VERBOSE
+    A_bcrs->describe(*out, Teuchos::VERB_EXTREME);
+#endif
+
+    // Create MultiVector
+    RCP<tpetra_multivector_type> X = Teuchos::rcp(new tpetra_multivector_type(A_bcrs->getDomainMap(), nrhs));
+    {
+      RCP<TimeMonitor> timerMultiVectorFill
+        = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("4) MultiVectorFill")));
+      
+      auto value = X->getLocalView<typename exec_space::memory_space>();
+      auto map = X->getMap()->getLocalMap();
+      Kokkos::parallel_for
+        (value.extent(0), KOKKOS_LAMBDA(const LO i) {
+          const GO gid = map.getGlobalElement(i);
+          for (LO j=0,jend=value.extent(1);j<jend;++j) 
+            value(i, j) = get_multi_vector_entry<value_type>(gid, j);
+        });
+    }
+    
+#if PRINT_VERBOSE
+    X->describe(*out, Teuchos::VERB_EXTREME);
+#endif
+
+    RCP<tpetra_multivector_type> B_bcrs = Teuchos::rcp(new tpetra_multivector_type(A_bcrs->getRangeMap(),  nrhs));
+
+    // matrix vector multiplication
+    {
+      for (LO iter=0;iter<repeat;++iter) {
+        RCP<TimeMonitor> timerBlockCrsApply 
+          = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("5) BlockCrs Apply")));        
+        A_bcrs->apply(*X, *B_bcrs);
+      }
+    }
+      
+    // direct conversion: block crs -> point crs 
+    RCP<tpetra_crs_matrix_type> A_crs;
+    {
+      RCP<TimeMonitor> timerConvertBlockCrsToPointCrs 
+        = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("6) Conversion from BlockCrs to PointCrs")));
+
+      // construct row mapd and column map
+      // row map can be obtained from A_bcrs->getDomainMap();
+      // we anyway need both and Tpetra does not provide an interface to contruct  
+      // crs matrix without column map
```

Comment:
What do you mean by "Tpetra does not provide an interface to construct crs matrix without column map"?  I think you mean "an interface that takes a Kokkos::StaticCrsGraph".  Could you please create a Tpetra issue for this missing feature and refer to it in this comment, so that once the feature exists, we can add it here?  Thanks!

Reply:
updated the comment. see Tpetra_CrsMatrix_decl.hpp, Line 504. It has a constructor with a local matrix, rowmap, colmap, domainmap, rangemap, and params. Essential inputs are local matrix and rowmap. However, when I attempt to use this constructor, it requires to provide a column map, which needs to be manually converted from block crs.